accepted/2021/winforms/streamline-application-bootstrap.md

```diff
@@ -0,0 +1,362 @@
+
+
+
+### How to resolve dpi settings?
+
+At this point an astute Windows Forms developer will say that there are currently 3 different ways to configure dpi settings (which may or may not be mutually exclusive or complimentary):
+* [app.manifest](https://docs.microsoft.com/windows/win32/sbscs/application-manifests) - Common Control related configuration, dealing with it was always a pain. Very likely that a significant part of Windows Forms developers have never heard of it.
+* [app.config ApplicationConfigurationSection](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/winforms/windows-forms-add-configuration-element) and [app.config AppContextSwitchOverrides](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element).
+* [`Application.SetHighDpiMode(HighDpiMode)`](https://docs.microsoft.com/dotnet/api/system.windows.forms.application.sethighdpimode).
+
+This proposal introduces a 4th way of configuring dpi, and to make it successful it has to (paraphrasing JRR Tolkien):
+
+
+> _One API to rule them all, One API to find them,_<br/>
+> _One API to bring them all and in the ApplicationConfiguration bind them..._
+
+
+The benefit the new approach provides is by facilitating the sharing of the information between the runtime and the designer during the development phase, as well as unifying how dpi settings are configured. This benefit is believed to outweigh the need to remove several lines of code from `Main()` method and adding several MSBuild properties.
+
+### Go further?
+
+At this point an astute reader may ask why no go further and offer an API similar to a ASP.<!-- -->NET Core bootstrap? Wouldn't it be nice to write something like:
+```cs
+[STAThread]
+static void Main()
+{
+    Application
+        .Configure(() =>
+        {
+            // application configuration logic
+        })
+        .Run(new Form1());
+}
+```
+
+Indeed, this sounds like a grand idea. However in reality Windows Forms app often have elaborate application bootstrap logic (e.g. [this](https://github.com/gitextensions/gitextensions/blob/master/GitExtensions/Program.cs), [this](https://github.com/luisolivos/Halconet/blob/4ed328a15ba69aab00e92076774977143790b5d3/HalcoNET%205.8.0/Ventas/Program.cs), or [this](https://github.com/aaraujo2019/PEQMIMERIA/blob/e11b6d57e3bbbe505834cbfeb717fe87066d6850/DBMETAL_SHARP/DBMETAL_SHARP/Program.cs)) that won't be possible in this scenario, and will require deeper investigations and feasibility studies.
+
+In the long term we could consider supporting a host/builder model, similar to the suggestion in [dotnet/winforms#2861](https://github.com/dotnet/winforms/issues/2861) and implementation in [alex-oswald/WindowsFormsLifetime](https://github.com/alex-oswald/WindowsFormsLifetime). E.g.:
+```cs
+[main: STAThread]
+CreateHostBuilder().Build().Run();
+
+public static IHostBuilder CreateHostBuilder() =>
```

Comment:
Why is this method definition here? It seems unnecessary.

Reply:
Agreed. It would be great if this was as streamlined as the latest [ASP.NET Core templates](https://github.com/dotnet/aspnetcore/blob/543eb19/src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs).